src/reducers/playground.ts

```diff
@@ -1,28 +1,106 @@
-import { Action, Reducer } from 'redux'
-import { IUpdateEditorValue, UPDATE_EDITOR_VALUE } from '../actions/playground'
+import { Reducer } from 'redux'
+import {
+  CLEAR_CONTEXT,
+  CLEAR_REPL_INPUT,
+  CLEAR_REPL_OUTPUT,
+  EVAL_INTERPRETER_ERROR,
+  EVAL_INTERPRETER_SUCCESS,
+  HANDLE_CONSOLE_LOG,
+  IAction,
+  SEND_REPL_INPUT_TO_OUTPUT,
+  UPDATE_EDITOR_VALUE,
+  UPDATE_REPL_VALUE
+} from '../actions/actionTypes'
+import { createContext } from '../slang'
+import { CodeOutput, defaultPlayground, InterpreterOutput, IPlaygroundState } from './states'
 
-export interface IPlaygroundState {
-  editorValue: string
-}
-
-/**
- * The default (initial) state of the `IPlaygroundState`
- */
-const defaultState: IPlaygroundState = {
-  editorValue: ''
-}
-
-/**
- * The reducer for `IPlaygroundState`
- *
- * UPDATE_EDITOR_VALUE: Update the `editorValue` property
- */
-export const reducer: Reducer<IPlaygroundState> = (state = defaultState, action: Action) => {
+export const reducer: Reducer<IPlaygroundState> = (state = defaultPlayground, action: IAction) => {
+  let newOutput: InterpreterOutput[]
+  let lastOutput: InterpreterOutput
   switch (action.type) {
     case UPDATE_EDITOR_VALUE:
       return {
         ...state,
-        editorValue: (action as IUpdateEditorValue).payload
+        editorValue: action.payload
+      }
+    case UPDATE_REPL_VALUE:
+      return {
+        ...state,
+        replValue: action.payload
+      }
+    case CLEAR_REPL_INPUT:
+      return {
+        ...state,
+        replValue: ''
+      }
+    case CLEAR_REPL_OUTPUT:
+      return {
+        ...state,
+        output: []
+      }
+    case CLEAR_CONTEXT:
+      return {
+        ...state,
+        context: createContext()
+      }
+    case HANDLE_CONSOLE_LOG:
+      /* Possible cases:
+       * (1) state.output === [], i.e. state.output[-1] === undefined
+       * (2) state.output[-1] is not RunningOutput
+       * (3) state.output[-1] is RunningOutput */
+      lastOutput = state.output.slice(-1)[0]
+      if (lastOutput === undefined || lastOutput.type !== 'running') {
+        newOutput = state.output.concat({
+          type: 'running',
+          consoleLogs: [action.payload]
+        })
+      } else {
+        lastOutput.consoleLogs = lastOutput.consoleLogs.concat(action.payload)
```

Comment:
this mutates `oldstate.output[oldstate.output-1].consoleLogs`, which in Redux is an antipattern